src/bank.cpp

```diff
@@ -0,0 +1,72 @@
+#include "openmc/bank.h"
+
+#include "openmc/capi.h"
+#include "openmc/error.h"
+
+#include <cstdint>
+
+// Explicit template instantiation definition
+template class std::vector<openmc::Particle::Bank>;
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+namespace simulation {
+
+std::vector<Particle::Bank> source_bank;
+std::vector<Particle::Bank> fission_bank;
+std::vector<Particle::Bank> secondary_bank;
+#ifdef _OPENMP
+std::vector<Particle::Bank> master_fission_bank;
+#endif
+
+} // namespace simulation
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+void free_memory_bank()
+{
+  simulation::source_bank.clear();
+  #pragma omp parallel
+  {
+    simulation::fission_bank.clear();
+  }
+#ifdef _OPENMP
+  simulation::master_fission_bank.clear();
+#endif
+}
+
+//==============================================================================
+// C API
+//==============================================================================
+
+extern "C" int openmc_source_bank(void** ptr, int64_t* n)
+{
+  if (simulation::source_bank.size() == 0) {
```

Comment:
I'll mention it only here, but are NULL-checks missing deliberately? There are reasons that could _arguably_ mean that they aren't required, but I can't find any references in the source code.